src/main/java/org/wildfly/security/credential/external/ExternalCredential.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2015 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.security.credential.external;
+
+import org.wildfly.common.Assert;
+import org.wildfly.security.credential.Credential;
+
+import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
+import java.security.Provider;
+import java.security.Security;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * External credential.
+ * It is source for credentials obtained from external source.
+ *
+ * @author <a href="mailto:pskopek@redhat.com">Peter Skopek</a>
+ */
+public class ExternalCredential {
```

Comment:
It is possible to use CallbackHandlers the way you described. I used Providers as it seems me to be easier for users. 
They can:
1. create their own ExternalCredentialSpi impl. and pack it in together with their CreadentialStoreSpi implementation 
2. use Elytron ExternalCredential providers with their implementation of CredentialStoreSpi 
3. implement ExternalCredentialSpi and use it with our CredentialStoreSpi impl.
In all case they will deploy their module and register it as another security provider.
In case of their own callback handler used with our credential store they will have to pass module name and class name to it. 
I used three different providers because there are three ways currently supported in EAP 6. (CLASS, EXT and CMD).